File: apps/web/lib/validations/kanban.ts
/**
 * Kanban Validation Schemas - Sprint 15 Phase B
 *
 * Zod schemas for validating kanban-related API requests.
 * Provides runtime validation and TypeScript type inference.
 *
 * @see types/kanban.ts for TypeScript interfaces
 * @see lib/constants/status/ticket.ts for status constants
 */

import { z } from 'zod';
import { TICKET_STATUS_VALUES, type TicketStatus } from '@/lib/constants/status';

// ============================================================================
// CONSTANTS
// ============================================================================

const MAX_MOVES_PER_REQUEST = 100; // Reasonable limit for bulk operations

// ============================================================================
// STATUS SCHEMA
// ============================================================================

/**
 * Valid ticket status values for kanban columns.
 * Uses the status constants as the source of truth.
 */
export const KanbanStatusSchema = z.enum(TICKET_STATUS_VALUES as [string, ...string[]]) as z.ZodType<TicketStatus>;

// ============================================================================
// MOVE TICKET SCHEMAS
// ============================================================================

/**
 * Schema for moving a single ticket to a new column/position.
 * Used by PATCH /api/tickets/[id]/move
 */
export const MoveTicketSchema = z.object({
  /**
   * Target status (kanban column) for the ticket.
   * Must be one of: backlog, todo, in-progress, in-review, done
   */
  status: KanbanStatusSchema,

  /**
   * Target position within the column (0-indexed).
   * 0 = top of column, higher values = lower in column.
   * If greater than column length, ticket is placed at bottom.
   */
  displayOrder: z
    .number()
    .int('Display order must be an integer')
    .nonnegative('Display order cannot be negative')
    .max(10000, 'Display order cannot exceed 10,000'),
});

export type MoveTicketInput = z.infer<typeof MoveTicketSchema>;

/**
 * Schema for a single move operation in bulk reorder.
 */
export const BulkMoveItemSchema = z.object({
  ticketId: z
    .number()
    .int('Ticket ID must be an integer')
    .positive('Ticket ID must be positive'),
  status: KanbanStatusSchema,
  displayOrder: z
    .number()
    .int('Display order must be an integer')
    .nonnegative('Display order cannot be negative')
    .max(10000, 'Display order cannot exceed 10,000'),
});

export type BulkMoveItemInput = z.infer<typeof BulkMoveItemSchema>;

/**
 * Schema for bulk reorder request.
 * Used by PATCH /api/tickets/reorder
 */
export const BulkReorderSchema = z.object({
  moves: z
    .array(BulkMoveItemSchema)
    .min(1, 'At least one move is required')
    .max(MAX_MOVES_PER_REQUEST, `Cannot exceed ${MAX_MOVES_PER_REQUEST} moves per request`),
});

export type BulkReorderInput = z.infer<typeof BulkReorderSchema>;

// ============================================================================
// QUERY PARAMETER SCHEMAS
// ============================================================================

/**
 * Schema for kanban board query parameters.
 * Used by GET /api/sprints/[sprintId]/kanban
 */
export const KanbanBoardQuerySchema = z.object({
  /**
   * Include ghost cards for parent/child relationships.
   * Defaults to true for full kanban experience.
   */
  includeGhosts: z
    .string()
    .transform((val) => val === 'true')
    .default('true')
    .optional(),

  /**
   * Filter by ticket kind (feature, task, bug, etc.)
   */
  kind: z
    .string()
    .optional(),

  /**
   * Filter by assignee
   */
  assignee: z
    .string()
    .optional(),
});

export type KanbanBoardQueryInput = z.infer<typeof KanbanBoardQuerySchema>;

/**
 * Schema for sprint ID parameter (cuid format).
 */
export const SprintIdParamSchema = z.object({
  sprintId: z.string().min(1, 'Sprint ID is required'),
});

export type SprintIdParamInput = z.infer<typeof SprintIdParamSchema>;

// ============================================================================
// ROADMAP OVERVIEW SCHEMAS
// ============================================================================

/**
 * Schema for roadmap overview query parameters.
 * Used by GET /api/roadmap/overview
 */
export const RoadmapOverviewQuerySchema = z.object({
  /**
   * Project ID (required for multi-tenancy)
   */
  projectId: z
    .string()
    .transform((val) => parseInt(val, 10))
    .refine((val) => !isNaN(val) && val > 0, 'Project ID must be a positive integer'),

  /**
   * Include ticket counts per sprint.
   * Defaults to true for dashboard views.
   */
  includeCounts: z
    .string()
    .transform((val) => val === 'true')
    .default('true')
    .optional(),
});

export type RoadmapOverviewQueryInput = z.infer<typeof RoadmapOverviewQuerySchema>;

// ============================================================================
// PROGRESS UPDATE SCHEMAS
// ============================================================================

/**
 * Schema for progress cascade response.
 * Returned after status changes that affect progress.
 */
export const ProgressCascadeSchema = z.object({
  ticketId: z.number().int().positive(),
  previousStatus: KanbanStatusSchema.optional(),
  newStatus: KanbanStatusSchema.optional(),
  parentProgress: z.number().min(0).max(100).optional(),
  sprintProgress: z.number().min(0).max(100).optional(),
  phaseProgress: z.number().min(0).max(100).optional(),
});

export type ProgressCascadeOutput = z.infer<typeof ProgressCascadeSchema>;

File: apps/mcp-server-python/src/tools/kanban.py
"""
Kanban tools: board view and ticket movement between columns.

Profile: core (always loaded)

API endpoints:
- GET /api/sprints/{sprintId}/kanban — get board with columns
- PATCH /api/tickets/{id}/move — move ticket between columns
"""

from typing import Literal

from src.tools._base import get_client, build_success, build_error, resolve_ticket_id
from src.logger import get_logger

logger = get_logger("tools.kanban")


async def projectpulse_kanban_getBoard(
    sprintId: str,
) -> str:
    """Get complete kanban board for a sprint with tickets grouped by column.

    Returns the 5-column board (backlog, todo, in-progress, in-review, done)
    with tickets, ghost cards, and statistics.

    Get sprintId from sprint_getCurrentPosition or roadmap_getPhaseProgress.

    Args:
        sprintId: Sprint ID (CUID string)
    """
    try:
        client = get_client()
        data = await client.get(f"/sprints/{sprintId}/kanban")
        if "error" in data:
            return build_error(data.get("message", data["error"]))
        return build_success(data)

    except Exception as e:
        logger.error("kanban_getBoard failed", error=str(e), sprintId=sprintId)
        return build_error(f"Failed to get kanban board: {e}")


async def projectpulse_kanban_moveTicket(
    status: Literal["backlog", "todo", "in-progress", "in-review", "done"],
    displayOrder: int,
    ticketId: int | None = None,
    ticketNumber: int | None = None,
    projectId: int | None = None,
) -> str:
    """Move a ticket between kanban columns with automatic progress cascade.

    When a ticket moves, progress auto-propagates up the hierarchy:
    ticket → sprint → phase. Moving to 'done' increases parent progress.

    Args:
        status: Target column — backlog, todo, in-progress, in-review, done
        displayOrder: Position within the column (0-10000)
        ticketId: Global ticket ID
        ticketNumber: Project-scoped number (e.g., #5)
        projectId: Required with ticketNumber (auto-fills from auth)
    """
    try:
        tid = await resolve_ticket_id(ticketId, ticketNumber, projectId)
        client = get_client()
        data = await client.patch(f"/tickets/{tid}/move", json={
            "status": status,
            "displayOrder": displayOrder,
        })
        if "error" in data:
            return build_error(data.get("message", data["error"]))
        return build_success(data)

    except Exception as e:
        logger.error("kanban_moveTicket failed", error=str(e))
        return build_error(f"Failed to move ticket: {e}")
